massa-ledger/src/ledger_db.rs

```diff
@@ -0,0 +1,279 @@
+// Copyright (c) 2022 MASSA LABS <info@massa.net>
+
+use massa_hash::Hash;
+use massa_models::{Address, Amount, DeserializeVarInt, SerializeVarInt};
+use rocksdb::{
+    ColumnFamilyDescriptor, Direction, IteratorMode, Options, ReadOptions, WriteBatch, DB,
+};
+use std::{collections::BTreeMap, str::FromStr};
+
+use crate::{ledger_changes::LedgerEntryUpdate, LedgerEntry, SetOrDelete, SetOrKeep};
+
+// TODO: remove rocks_db dir when sled is cut out
+const DB_PATH: &str = "../massa-node/storage/ledger/rocks_db";
+const LEDGER_CF: &str = "ledger";
+const METADATA_CF: &str = "metadata";
+const OPEN_ERROR: &str = "critical: rocksdb open operation failed";
+const CRUD_ERROR: &str = "critical: rocksdb crud operation failed";
+const CF_ERROR: &str = "critical: rocksdb column family operation failed";
+
+/// Ledger sub entry enum
+pub enum LedgerSubEntry {
+    /// Balance
+    Balance,
+    /// Bytecode
+    Bytecode,
+    /// Datastore entry
+    Datastore(Hash),
+}
+
+pub(crate) struct LedgerDB(pub(crate) DB);
+
+/// Destroy the disk ledger db and free the lock
+pub fn destroy_ledger_db() {
+    DB::destroy(&Options::default(), DB_PATH).expect(OPEN_ERROR);
+}
+
+macro_rules! balance_key {
+    ($addr:ident) => {
+        format!("1:{}", $addr).as_bytes()
+    };
+}
+
+// NOTE: still handle separate bytecode for now to avoid too many refactoring at once
+macro_rules! bytecode_key {
+    ($addr:ident) => {
+        format!("2:{}", $addr).as_bytes()
+    };
+}
+
+macro_rules! data_key {
+    ($addr:ident, $key:ident) => {
+        format!("{}:{}", $addr, $key).as_bytes()
+    };
+}
+
+macro_rules! data_prefix {
+    ($addr:ident) => {
+        format!("{}:", $addr).as_bytes()
+    };
+}
+
+pub fn end_prefix(prefix: &[u8]) -> Option<Vec<u8>> {
+    let mut end_range = prefix.to_vec();
+    while let Some(0xff) = end_range.last() {
+        end_range.pop();
+    }
+    if let Some(byte) = end_range.last_mut() {
+        *byte += 1;
+        Some(end_range)
+    } else {
+        None
+    }
+}
+
+impl LedgerDB {
+    pub fn new() -> Self {
+        // db options
+        let mut db_opts = Options::default();
+        db_opts.create_if_missing(true);
+        db_opts.create_missing_column_families(true);
+
+        // database init
+        let db = DB::open_cf_descriptors(
+            &db_opts,
+            DB_PATH,
+            vec![
+                ColumnFamilyDescriptor::new(LEDGER_CF, Options::default()),
+                ColumnFamilyDescriptor::new(METADATA_CF, Options::default()),
+            ],
+        )
+        .expect(OPEN_ERROR);
+
+        // return database
+        LedgerDB(db)
+    }
+
+    pub fn put(&mut self, addr: &Address, ledger_entry: LedgerEntry) {
+        let mut batch = WriteBatch::default();
+        let handle = self.0.cf_handle(LEDGER_CF).expect(CF_ERROR);
+
+        // balance
+        batch.put_cf(
+            handle,
+            balance_key!(addr),
+            ledger_entry.parallel_balance.to_raw().to_varint_bytes(),
```

Comment:
we *always* need to avoid usig to_raw. Here Balance::to_bytes_compact should be used

Reply:
the day we change the underlying representation of Amount, we will have to find all the places where we made assumptions instead of using the functions exposed by the Amount module whose role is exactly that.

For the Result, you might just unwrap() it for now with a comment. In the future we will stop using this trait after the work on serialization improvements :)

Reply:
Done.